### ISSUE TITLE
About the input format of pretraining

### ISSUE THREAD

**Issue by mini-ops996** [macwiatrak/Bacformer#8]
Thank you for the excellent open-source project. I have another question: Could you provide an example of the data input or a code example for the preprocessing during the training phase?

**Comment by macwiatrak** [comment]
Thanks for raising the issue. I will upload a small dataset sample in the next few days which shows the input format tested with the pretraining script.

**Comment by zhaorui-bi** [comment]
Could you provide the pretraining script for Bacformer?

**Comment by macwiatrak** [comment]
@mini-ops996 I have added a toy pretraining script, including a data sample, see PR https://github.com/macwiatrak/Bacformer/pull/10.

Here is the script to run the toy example:
```python
python bacformer/modeling/run_pretraining.py \
	--input-dir /files/pretraining_data_sample \
	--output-dir <your-output-dir> \
	--batch-size 2 \
	--gradient-accumulation-steps 1 \
	--logging-steps 1 \
	--eval-steps 20 \
	--save-steps 100 \
	--warmup-proportion 0.0
```
As there are only 3 samples, make sure to keep `gradient-accumulation-steps=1`.

The data sample is the input format to Bacformer pretraining, the `protein_embeddings` column are protein sequences embedded with https://huggingface.co/facebook/esm2_t12_35M_UR50D model, ordered by their location on the genome. The `contig_idx` columns is the `contig_id` to which each protein belongs to and the `prot_cluster_id` is the result of unsupervised `kmeans` clustering of protein embeddings used as a pretraining label for masked protein embeddings.

Let me know if anything was unclear and apologies for the delay! I will merge the MR once you confirm things work.

@zhaorui-bi I now changed the [argparser.py](https://github.com/macwiatrak/Bacformer/blob/feat-add-pretraining-example/bacformer/modeling/argparser.py) file to have the default params used for pretraining, but pasting the full pretraining script below as well for the Bacformer with masked objective on 4 GPUs.

```python
torchrun --nproc_per_node=4 bacformer/modeling/run_pretraining.py \
	--input-dir <input-dir> \
	--output-dir <output-dir> \
	--batch-size 1 \
	--logging-steps 100 \
	--warmup-proportion 0.1 \
	--mgm-probability 0.15 \
	--lr 0.00015 \
	--max-n-proteins 6000 \
	--num-hidden-layers 12 \
        --num-attention-heads 8 \
        --hidden-size 480 \
        --n-total-samples 1203731 \
        --gradient-accumulation-steps 8 \
        --dataloader-num-workers 10 \
        --max-epochs 10 \
        --eval-steps 8000 \
        --save-steps 8000
```

We did not release the full dataset as it takes >3TB, but the input was a set of files (i.e. partitions) with each file having ~50 unique genomes.

Let me know if anything was not clear!

**Comment by macwiatrak** [comment]
@oooo111  to continue pretraining, you also need the labels. For Bacformer the labels were computed by doing a nearest neighbor search from each ESM-2 embedding to 50k cluster centroids (computed by performing k-means on a large set of ESM-2 embeddings).

To make it easier for you, I have uploaded the 50k cluster centroids to Google Drive - here is the link https://drive.google.com/file/d/1oAIoKLOD0aRVST9TiMaSR_OCqjPScBT-/view?usp=sharing (<100 MB).

For the nearest neighbor search I recommend using the [cuML](https://github.com/rapidsai/cuml) package. Here is a short snippet on how one can use it. I recommend doing it as a preprocessing step (although it's very fast so can be done during training as well).

```python
import numpy as np

from cuml.neighbors import NearestNeighbors  # this only works with a NVIDIA GPU

# load protein embeddings, make sure to exclude any special tokens beforehand
protein_embeddings = np.load(<protein-embeddings-file>)
cluster_centers = np.load(<cluster_centers_file_path>)

# preparing the model
model = NearestNeighbors(n_neighbors=1)
model.fit(cluster_centers)

# get 1 nearest neighbor
dists, indices = model.kneighbors(protein_embeddings)

# reshape for the array to be one dimensional
labels = indices.reshape(-1)
```

**Comment by csadorf** [comment]
@macwiatrak I'm glad cuml is useful for your application. Please let us know if you have any feedback for us!

**Comment by macwiatrak** [comment]
Thanks @csadorf! I love cuML! I will send you an email with feedback after I am back from my annual leave.

**Comment by oooo111** [comment]
@macwiatrak

 Dear Author,
I am currently conducting research on the six ESKAPE multidrug-resistant bacteria, and I would like to evaluate performance at the operon level. Could you please provide me with a complete operon prediction dataset, restricted to the ESKAPE species only? I would greatly appreciate your assistance.

Best regards,
